package_modules/apt_get

```diff
@@ -0,0 +1,245 @@
+    sys.stdin.readlines()
+
+    if online:
+        result = subprocess.call([apt_get_cmd] + apt_get_options + ["update"], stdout=sys.stderr)
+        if result != 0:
+            return result
+
+    process = subprocess.Popen([apt_get_cmd] + apt_get_options + ["-s", "upgrade"], stdout=subprocess.PIPE)
+    for line in process.stdout:
+        # Example of lines that we try to match:
+        #        (name)  (old version (ignored))  (new version)     (repository(ies) (ignored))  (arch)
+        #           |              |                    |                        |                 |
+        #           V              V                    V                        V                 V
+        #   Inst php5-cli [5.3.10-1ubuntu3.17] (5.3.10-1ubuntu3.18 Ubuntu:12.04/precise-updates [amd64]) []
+        #
+        # Note architecture included in the name on this one:
+        #   Inst php5-cli:i386 [5.3.10-1ubuntu3.17] (5.3.10-1ubuntu3.18 Ubuntu:12.04/precise-updates [i386]) []
+        #
+        # Note multiple repositories in this one:
+        #   Inst linux-libc-dev [2.6.32-48squeeze4] (2.6.32-48squeeze6 Debian:6.0.10/oldstable, Debian-Security:6.0/oldoldstable [amd64])
+        #
+        #                               name                   old version       new version
+        #                                 |                         |                 |
+        #                         /-------+-------\              /--+--\       /------+-------\
+        match = re.match("^Inst\s+(?P<name>[^\s:]+)(?::\S+)?\s+\[[^]\s]+\]\s+\((?P<version>\S+)" +
+
+        #                         repository(ies)              arch
+        #                               |                        |
+        #                    /----------+----------\     /-------+-------\
+                         "\s+[^,\s]+(?:,\s+[^,\s]+)*\s+\[(?P<arch>[^]\s]+)\]\).*", line)
+
+        if match != None:
+            sys.stdout.write("Name=" + match.group("name") + "\n")
+            sys.stdout.write("Version=" + match.group("version") + "\n")
+            sys.stdout.write("Architecture=" + match.group("arch") + "\n")
+
+    return 0
+
+
+def one_package_argument(name, arch, version, is_apt_install):
+    args = []
+    archs = []
+    if arch:
+        archs.append(arch)
+    elif is_apt_install:
+        # Here we have no specified architecture and we are installing.
+        # If we have existing versions, operate on those, instead of the
+        # platform default.
+        process = subprocess.Popen([dpkg_query_cmd, "--showformat", "${Architecture}=${Status}\n",
+                                    "-W", name + ":*"], stdout=subprocess.PIPE)
+        for line in process.stdout:
+            # The space before "installed" is important, because it can be "not-installed".
+            if "=" in line:
+                arch, stat = line.split("=", 1)
+                if stat.find(" installed") >= 0:
+                    archs.append(arch)
```

Comment:
For reference, rewriting this as a list comprehension would look like:
```
archs = [arch for arch, stat in (line.split('=', 1) for line in process.stdout if '=' in line) if stat.find('installed) >= 0]
```
Notice that the inner generator (iterating over line) is in simple parentheses; this is just a generator, so we don't gather up its results into a list (or a tuple, as you might have expected (...) to give you), you'll just iterate them in the course of building your list of archs.

I'm used to generators, so find the above reads quite plainly; but I can quite understand that you might prefer the iterative form if you find that clearer ...